src/main/scala/temple/builder/DatabaseBuilder.scala

```diff
@@ -0,0 +1,57 @@
+package temple.builder
+
+import temple.DSL.semantics.{Annotation, Attribute, AttributeType, ServiceBlock}
+import temple.generate.database.ast.ColumnConstraint.Check
+import temple.generate.database.ast._
+
+/** Construct database queries from a Templefile structure */
+object DatabaseBuilder {
+
+  private def generateMaxMinConstraints[T](name: String, max: Option[T], min: Option[T]): Seq[ColumnConstraint] = {
+    val maxCondition = max.map(v => Check(name, ComparisonOperator.LessEqual, v.toString))
+    val minCondition = min.map(v => Check(name, ComparisonOperator.GreaterEqual, v.toString))
+    Seq(maxCondition, minCondition).flatten
+  }
+
+  private def toColDef(name: String, attribute: Attribute): ColumnDef = {
+    val valueConstraints = attribute.valueAnnotations.map {
+      case Annotation.Unique => ColumnConstraint.Unique
+    }.toSeq
+
+    val (colType, constraints) = attribute.attributeType match {
+      case AttributeType.BoolType     => (ColType.BoolCol, valueConstraints)
+      case AttributeType.DateType     => (ColType.DateCol, valueConstraints)
+      case AttributeType.DateTimeType => (ColType.DateTimeTzCol, valueConstraints)
+      case AttributeType.TimeType     => (ColType.TimeCol, valueConstraints)
+      case AttributeType.ForeignKey   => (ColType.IntCol(4), valueConstraints)
+      case AttributeType.BlobType(_)  => (ColType.BlobCol, valueConstraints)
+      case AttributeType.IntType(max, min, precision) =>
+        (ColType.IntCol(precision), valueConstraints ++ generateMaxMinConstraints(name, max, min))
+      case AttributeType.FloatType(max, min, precision) =>
+        (ColType.FloatCol(precision), valueConstraints ++ generateMaxMinConstraints(name, max, min))
+      case AttributeType.StringType(max, min) =>
+        val colType = if (max.isDefined) ColType.BoundedStringCol(max.get) else ColType.StringCol
```

Comment:
This feels safer than using `.get`:
`max.fold(ColType.StringCol) { ColType.BoundedStringCol(_) }`

Reply:
I'm getting a compiler error if I try this, even when I annotate the type of `colType`

Reply:
As an aside, the implementation of `fold` is exactly what I wrote originally - checks if it's empty and uses `.get` otherwise

Reply:
Never mind then